promql/functions.go

```diff
@@ -79,22 +75,29 @@ func funcDelta(ev *evaluator, args Expressions) model.Value {
 		}
 		resultValue := lastValue - samples.Values[0].Value + counterCorrection
 
-		targetInterval := args[0].(*MatrixSelector).Range
-		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp)
-		if sampledInterval == 0 {
-			// Only found one sample. Cannot compute a rate from this.
-			continue
+		// Duration between first/last samples and boundary of range.
+		durationToStart := samples.Values[0].Timestamp.Sub(rangeStart).Seconds()
+		durationToEnd := rangeEnd.Sub(samples.Values[len(samples.Values)-1].Timestamp).Seconds()
+
+		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp).Seconds()
+		averageDurationBetweenSamples := sampledInterval / float64(len(samples.Values))
```

Comment:
I'm currently more worried that we have to explain to the users how it changes. I don't want to explain a trade-off of computational cost and why we took a way we know is sub-optimal because of that. If we are really worried about the cost, we should benchmark it now and then come to a conclusion for good instead of changing the behavior later again.

The biggest jump in cost will anyway happen, it is for long-range functions that didn't need to take counter resets into account so far. Here again, providing an explicitly non-extrapolating function would allow for a low-cost function. With this PR, we always have to go through all samples, even without counter reset awareness.